### PR TITLE
feat(angular): add an option to add strict type checking

### LINKED ISSUE #3383: Angular Strict Mode Support
## Description

When generating a new application using the Angular CLI we can pass `--strict` flag to turn on strict template type checking and strict TypeScript checking.

It would be good to have this option available when using the Nx Angular Application and Library schematics too.

For example, this would add the following in an Angular CLI application:

**tsconfig.base.json**

```json
/* To learn more about this file see: https://angular.io/config/tsconfig. */
{
  "compileOnSave": false,
  "compilerOptions": {
    "baseUrl": "./",
    "outDir": "./dist/out-tsc",
    "forceConsistentCasingInFileNames": true,
    "strict": true, // enable strict type checking
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "sourceMap": true,
    "declaration": false,
    "downlevelIteration": true,
    "experimentalDecorators": true,
    "moduleResolution": "node",
    "importHelpers": true,
    "target": "es2015",
    "module": "es2020",
    "lib": [
      "es2018",
      "dom"
    ]
  },
  "angularCompilerOptions": { // Add strict template checks
    "strictInjectionParameters": true,
    "strictTemplates": true
  }
}
```

## Motivation

Stricter type safety can be very useful especially in large enterprise applications. It would also be good to have parity with the @angular/cli generate schematic for this feature.

## Suggested Implementation

```
ng generate @nrwl/angular:application my-app --strict
```

## Alternate Implementations

Library support as well perhaps?
```
ng generate @nrwl/angular:library my-lib --strict
```

Thanks 😊

### PR DESCRIPTION
**Pull request by ashley-hunter**
## Current Behavior

Currently the Angular CLI provides the option to enable `strict` mode giving us stricter type checking in both TypeScript and Angular templates. However Nx does not currently have equivalent functionality.

## Expected Behavior

Adding support for strict type checking to Angular application and library generate schematics.

Performs the following configuration changes:

- Enables strict mode in TypeScript, as well as other strictness flags recommended by the TypeScript team. Specifically, `forceConsistentCasingInFileNames`, `noImplicitReturns`, `noFallthroughCasesInSwitch`.

- Turns on strict Angular compiler flags `strictTemplates` and `strictInjectionParameters`.

These match the flags used in the standard CLI strict mode.

E.g.

`nx generate application myapp --strict`
`nx generate lib mylib --strict`

## Related Issue(s)

Fixes #3383

### CODE REVIEW

**Comment by brandonroberts** [comment]
Thanks for the PR @ashley-hunter. Why do we need to update the tsconfig.json files manually instead of just forwarding the `strict` flag to the `@schematics/angular:application` schematic?

**Comment by ashley-hunter** [comment]
Hi @brandonroberts, thanks for your feedback! Yes, that was the first approach I tried, however at least from what I can tell, the tsconfig that is produced by the `@schematics/angular:application` schematic is overridden. I see we have a few possible things that replace/delete the tsconfig produced by the application schematic:

1. The schematic files folder contains a tsconfig.json template file
2. There is a step to update the json content of the file:

```typescript
// TODO: Remove this after Angular 10.1.0
updateJsonInTree('tsconfig.json', () => ({
    files: [],
    include: [],
    references: []
})),
```

3. There is a step to delete the tsconfig.json file:
```typescript
// TODO: Remove this after Angular 10.1.0
(host) => {
    host.delete('tsconfig.json');
},
```

However, if you have any suggestions on how to workaround this let me know and I can make the changes.

Thank you

**Comment by brandonroberts** [comment]
@ashley-hunter you could modify the template for the `packages/angular/src/schematics/application/files/tsconfig.json` based on strict option, and update the `tsconfig.e2e.json` also, but after looking at it also there is a cleaner separation of the updates with what you have, so I think we're good here.

File: packages/angular/src/schematics/library/lib/enable-strict-type-checking.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  // set the default so future libraries will default to strict mode
+  // unless the user has previously set this to false by default
+  return updateWorkspace((workspace) => {
+    workspace.extensions.schematics = workspace.extensions.schematics || {};
+
+    workspace.extensions.schematics['@nrwl/angular:library'] =
+      workspace.extensions.schematics['@nrwl/angular:library'] || {};
+
+    workspace.extensions.schematics['@nrwl/angular:library'].strict =
+      workspace.extensions.schematics['@nrwl/angular:library'].strict ?? true;
```
Comment (brandonroberts) [inline packages/angular/src/schematics/library/lib/enable-strict-type-checking.ts:48]: Should this be `||`?

**Review by brandonroberts — CHANGES_REQUESTED** [review]

File: packages/angular/src/schematics/library/lib/enable-strict-type-checking.ts (hunk 1 above)
Comment (ashley-hunter) [reply packages/angular/src/schematics/library/lib/enable-strict-type-checking.ts:48]: If the user has previously set the `strict` value explicitly to be `false`, then using `||` would change it to `true` overriding their previously defined value, whereas using `??` would only set it to `true` if they had no previously defined value. 

I didn't think it was a good idea to override an option that a user had specifically defined?

**Review by brandonroberts — APPROVED** [review]

**Pull request merged**